src/FTPObject.jl

```diff
@@ -132,168 +132,156 @@ end
+"remote_path".
+
+If "remote_path" is a full file name path, then the file will be uploaded to the FTP
+using the full file path name. If "remote_path" is a path to a directory (which means
+it ends in "/"), then the file will be uploaded to the specified directory but with the
+"local_path" basename as the file name.
+
+# Arguments
+`ftp::FTP`: The FTP to deliver to. See FTPClient.FTP for details.
+`local_path::AbstractString`: The file path to the file we want to deliver.
+`remote_path::AbstractString`: The file/dir path that we want to deliver to.
+
+# Keywords
+`ftp_options=ftp.ctxt`: FTP Options
+`mode::FTP_MODE=binary_mode`: Set the ftp mode.
+`verbose=nothing`: Set the verbosity
 
-Upload IO object "local_file" to the FTP server and save as "remote_name".
+# Returns
+`Bool`: Returns a boolean with true if the file was successfully transfered, else false.
 """
 function upload(
     ftp::FTP,
-    local_file::IO,
-    remote_name::AbstractString;
+    local_path::AbstractString,
+    remote_path::AbstractString;
+    ftp_options=ftp.ctxt,
     mode::FTP_MODE=binary_mode,
-    verbose=nothing,
-)
+    verbose=nothing
+)::Bool
     _dep_verbose_kw(verbose, typeof(ftp), :upload)
 
-    try
-        ftp_put(ftp.ctxt, remote_name, local_file; mode=mode, verbose=verbose)
-    catch err
-        if isa(err, FTPClientError)
-            err.msg = "Failed to upload $remote_name."
-        end
-        rethrow()
+    # Whether or not the current file was successfully delivered to the FTP
+    success = false
+
+    # The location we are going to drop the file in the FTP server
+    if basename(remote_path) == ""
+        # If the remote path was just a directory, then the full remote path should be
+        # that directory plus the basename of the local_path
+        server_location = joinpath(dirname(remote_path), basename(local_path))
+    else
+        # If the remote path is a full file path, then just use that
+        server_location = remote_path
     end
-    return nothing
-end
 
+    # ftp_put requires an io, so open the file
+    open(local_path) do local_path_io
+        success = upload(
+            ftp, local_path_io, server_location;
+            ftp_options=ftp_options, mode=mode, verbose=verbose
+        )
+    end
+    return success
+end
 
 """
     upload(
         ftp::FTP,
-        local_file_paths::Vector{<:AbstractString},
-        ftp_dir<:AbstractString;
-        retry_callback::Function=(count, options) -> (count < 4, options),
-        retry_wait_seconds::Integer = 5,
-    )
-
-Uploads the files specified in local_file_paths to the directory specifed by
-ftp_dir. The files will have the same names.
-
-By default, will try to deliver the files 4 times with a 5 second wait in between
-each failed attempt.
+        local_path::AbstractString;
+        ftp_options=ftp.ctxt,
+        mode::FTP_MODE=binary_mode,
+        verbose=nothing,
+    ) -> Bool
```

Comment:
I'm not sure we need this version of upload as we should also support `upload(ftp, local_path, ".")`

Reply:
I guess in the case of "." it would fail since dirname(x) returns nothing. 
Thus the file would be written to "." rather than joinpath(".", basename(local_path))
So potentially there's a special case there for '.', but I would think it should work otherwise.